### ISSUE TITLE
Insert org-brain links in body of file

### ISSUE THREAD

**Issue by namdnguyen** [Kungsgeten/org-brain#259]
I was wondering if there was an easy way to insert an org-mode file link in the body of a file, instead of just in the property drawer or file property. This function would allow for quicker, wiki-style (in terms of workflow) link insertion, while also preserving the org-brain hierarchy of parent, child, and friend relationships. 

The reason I would want to insert in-body links is that it allows me to contextualize related notes better. This issue is somewhat related to #87, but is a more active approach in comparison.

### Existing Workflow with `org-insert-link` and `org-brain-add-...`

For an example use-case with the existing functions, if I were writing a new note and wanted to an org-mode link to another org-brain note at a specific point in my new note, I would have to go through a few steps:

1. Search for org-brain note to be inserted and open it.
2. Capture the file link with `org-store-link`.
3. Navigate back to new note.
4. Insert file link at the location I want it in the body with `org-insert-link`.
5. Since I want the relationship stored in org-brain, I would then use something like `org-brain-add-friendship`.

If I wanted to insert a link in the body of a new note for a note that doesn't yet exist, I would need to do `org-brain-add-friendship` first, then go through the rest to capture the file link and insert it.

### Existing Workflow with `org-brain-insert-relationships`

I also saw the `org-brain-insert-relationships` function, which inserts org-mode file links with the note titles as the link text. There are a couple things that add friction to the workflow with this method, too:

1. Still need to go through `org-brain-add-...` first to populate the results of `org-brain-insert-relationships`.
2. I would need to copy the link and insert it where I want and delete the unused results of `org-brain-insert-relationships`.
3. Link description editing needed if I want to shorten the link text.

### Proposed Workflow and Function

Something like `org-brain-insert-link` could wrap around `org-brain-add-...` and `org-insert-link`. `org-insert-link` is nice because it allows for editing the default link location and description interactively.

Here are some thoughts on how `org-brain-insert-link` could work:

1. `M-x org-brain-insert-link` pulls up search list like `org-brain-goto`.
2. If note already exist, prompt to select type of relationship with new note. If note does not yet exist, prompt to select type, then call appropriate `org-brain-add-...` function.
3. Pass file location and title to `org-create-link` (if possible), which allows for interactive editing of link and link description.

There could be an `org-brain-insert-link-local` function to generate the search list with just org-brain files. 

Let me know what you think, or if I'm not being clear!

**Comment by namdnguyen** [comment]
[Org-roam](https://org-roam.readthedocs.io/en/latest/tour/#inserting-links) provides a similar functionality, and I think other wiki-style notes packages probably provide something similar. 

The main thing that I like about org-brain is that distinctions provided by the different relationships. As I understand it, there is only one type of relationship in org-roam, which are bi-directional back links. I haven't had a chance to try out org-roam, yet so this might not be an accurate comparison.

A query that I have in mind is to sort the notes with the most friendships, which would automatically not include parents, children, or siblings, which is possible in org-brain but not org-roam.

**Comment by benjaminwd** [comment]
I'm curious if this is similar to #163 or if I'm not fully understanding the workflow? You can create links in the body of an entry using org-insert-link and the brain-<relationship> link type, which will then insert that link into the entries property drawer (as well as creating the link inline).

**Comment by namdnguyen** [comment]
I am not sure. I don't use brain-relationship links, so I don't know how they work. The workflow I have in mind is for rapid note creation where I can reference other notes, even if they don't exist yet. It's common to wikis and Roam. To insert links in text, I would like to avoid having to open the file to be referenced to visit it and store the link, go back to the new note to insert the link, and then create the brain relationship as a separate step with the new note. 

Wikis support creating backlinks to existing or not-yet-created pages with markup like with brackets like `[[note taking systems]]` or camel case `noteTakingSystem` or some other format. These work best for short names or with some kind of completion system. Since org-brain already supports selecting a note or creating a new one from a list provided by ivy or helm, this workflow could build on these functionalities.

If there is already a function that enables this workflow, could you point me to the name of the function? I can try it out and report back, if it covers this use-case.

**Comment by benjaminwd** [comment]
Ah, I see what you mean. org-brain will create a headline or file if you create a brain: link to it (for example, on an existing entry use org-brain-add-friendship and create the friendship for "Test" and it will create a new file or new top level headline for "Test", creating the relationship at the same time. It will do the same if you org-insert-link and create a brain-_relationship_: link). This works well for files, but I don't know how you would create anything below a top level headline. I'm not sure if it's quite your use case, but perhaps @Kungsgeten has a better idea.

**Comment by Kungsgeten** [comment]
Hi! From what I understand of the suggestion, I think @benjaminwd got it right.

When you're in an `org-mode` buffer (and the file is part of `org-brain`) you can use the normal `C-c C-l` keybinding (at least in vanilla `org-mode`) to run the `org-insert-link` command. This brings up a dialog of different link types. If you choose `brain-child` for instance you'll get a dialog where you can choose an existing `org-brain` entry (or create a new one). The chosen entry will then be added as a child of the entry you were editing. You can then click the link in order to goto the entry.

You can read more about this under _Editing from org-mode_ in the README.

**Comment by namdnguyen** [comment]
Thank you @benjaminwd and @Kungsgeten! I'll check out your suggestions and get back to you.

**Comment by namdnguyen** [comment]
Ok, I just tried it out. I did not know about the `org-insert-link` types because I didn't get how to use it when I read about it previously. Thank for the explanation. Along with your help and the README, I get them now, and they completely solve the use case I was asking about! I consider this issue to be closed.

The only suggestion I have is whether you think it would be a good idea to auto-populate the link description with the note title.
